go/store/nbs/archive_reader.go

```diff
@@ -0,0 +1,382 @@
+	chunkRefs []chunkRef
+	suffixes  []suffix
+	footer    footer
+}
+
+type chunkRef struct {
+	dict uint32
+	data uint32
+}
+
+type suffix [hash.SuffixLen]byte
+
+type footer struct {
+	indexSize     uint32
+	byteSpanCount uint32
+	chunkCount    uint32
+	metadataSize  uint32
+	dataCheckSum  sha512Sum
+	indexCheckSum sha512Sum
+	metaCheckSum  sha512Sum
+	formatVersion byte
+	fileSignature string
+	fileSize      uint64 // Not actually part of the footer, but necessary for calculating offsets.
+}
+
+// dataSpan returns the span of the data section of the archive. This is not generally used directly since we usually
+// read individual spans for each chunk.
+func (f footer) dataSpan() byteSpan {
+	return byteSpan{offset: 0, length: f.fileSize - archiveFooterSize - uint64(f.metadataSize) - uint64(f.indexSize)}
+}
+
+// totalIndexSpan returns the span of the entire index section of the archive. This span is not directly useful as
+// the index is broken into a compressed section and an uncompressed section. Use indexCompressedSpan and indexSuffixSpan
+func (f footer) totalIndexSpan() byteSpan {
+	return byteSpan{offset: f.fileSize - archiveFooterSize - uint64(f.metadataSize) - uint64(f.indexSize), length: uint64(f.indexSize)}
+}
+
+// indexCompressedSpan returns the span of the index section of the archive.
+func (f footer) indexCompressedSpan() byteSpan {
+	suffixLen := uint64(f.chunkCount * hash.SuffixLen)
+	totalIdx := f.totalIndexSpan()
+	return byteSpan{offset: totalIdx.offset, length: totalIdx.length - suffixLen}
+}
+
+func (f footer) indexSuffixSpan() byteSpan {
+	suffixLen := uint64(f.chunkCount * hash.SuffixLen)
+	totalIdx := f.totalIndexSpan()
+	compressedLen := totalIdx.length - suffixLen
+
+	return byteSpan{totalIdx.offset + compressedLen, suffixLen}
+}
+
+func (f footer) metadataSpan() byteSpan {
+	return byteSpan{offset: f.fileSize - archiveFooterSize - uint64(f.metadataSize), length: uint64(f.metadataSize)}
+}
+
+func newArchiveReader(reader io.ReaderAt, fileSize uint64) (archiveReader, error) {
+	footer, err := loadFooter(reader, fileSize)
+	if err != nil {
+		return archiveReader{}, err
+	}
+
+	indexSpan := footer.indexCompressedSpan()
+	secRdr := io.NewSectionReader(reader, int64(indexSpan.offset), int64(indexSpan.length))
+	rawReader := bufio.NewReader(secRdr)
+
+	errChan := make(chan error, 1)
+
+	redr, wrtr := io.Pipe()
+	go func() {
+		defer wrtr.Close()
+		err := gozstd.StreamDecompress(wrtr, rawReader)
+		if err != nil {
+			errChan <- errors.Wrap(err, "Failed to decompress archive index")
+		}
+
+		close(errChan)
+	}()
+	byteReader := bufio.NewReader(redr)
+
+	byteSpans := make([]byteSpan, footer.byteSpanCount+1)
+	byteSpans = append(byteSpans, byteSpan{offset: 0, length: 0}) // Null byteSpan to simplify logic.
+
+	offset := uint64(0)
+	for i := uint32(0); i < footer.byteSpanCount; i++ {
+		length := uint64(0)
+		err := binary.Read(byteReader, binary.BigEndian, &length)
+		if err != nil {
+			return archiveReader{}, err
```

Comment:
In all these early return cases, I think we might need to do something like `redr.Close()` so that the writing thread sees an error and doesn't block indefinitely. I think maybe we can just get away with a `defer redr.Close()` when we create it?